src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,760 @@
+use std::{
+    cell::UnsafeCell,
+    fmt::Debug,
+    pin::Pin,
+    sync::{atomic::AtomicU64, Arc},
+    task::{Context, Poll},
+};
+
+use bytes::BytesMut;
+use futures::{pin_mut, Future};
+use tokio::{
+    io::{AsyncRead, AsyncWriteExt, ReadBuf}, net::TcpStream, time::{Instant, Interval}
+};
+use tracing::debug;
+
+use crate::write_all_timeout;
+
+use super::{
+    common::{self, ClientTimeoutOptions},
+    error::RpcError,
+    message::{ReqType, RespType},
+    packet::{Decode, Encode, Packet, PacketsKeeper, ReqHeader, RespHeader, REQ_HEADER_SIZE},
+    utils::u64_to_usize,
+};
+
+/// TODO: combine `RpcClientConnectionInner` and `RpcClientConnection`
+struct RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    /// The TCP stream for the connection.
+    stream: UnsafeCell<TcpStream>,
+    /// Options for the timeout of the connection
+    timeout_options: ClientTimeoutOptions,
+    /// Stream auto increment sequence number, used to mark the request and response
+    seq: AtomicU64,
+    /// The instant of the connection, used to calculate the keep alive timeout
+    clock_instant: Instant,
+    /// Received keep alive timestamp, will mark the valid keep alive response.
+    /// If the keep alive response is not received in right mestamp, the connection will be closed,
+    /// If we receive other response, we will update the received_keepalive_timestamp too, just treat it as a keep alive response.
+    received_keepalive_timestamp: AtomicU64,
+    /// Send packet task
+    packets_keeper: PacketsKeeper<P>,
+    /// Response buffer, try to reuse same buffer to reduce memory allocation
+    /// In case of the response is too large, we need to consider the buffer size
+    /// Init size is 4MB
+    /// Besides, we want to reduce memory copy, and read/write the response to the same data buffer
+    resp_buf: UnsafeCell<BytesMut>,
+    /// Request encode buffer
+    req_buf: UnsafeCell<BytesMut>,
+    /// The Client ID for the connection
+    client_id: u64,
+    /// Is closed flag sender
+    is_closed_sender: flume::Sender<()>,
+    /// Is closed flag receiver
+    is_closed_receiver: flume::Receiver<()>,
+}
+
+// TODO: Add markable id for this client
+impl<P> Debug for RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RpcClientConnectionInner")
+            .field("client id", &self.client_id)
+            .finish_non_exhaustive()
+    }
+}
+
+/// Drop the client connection, send the close signal to the client
+/// Other side will receive the close signal and close the connection
+impl<P> Drop for RpcClientConnectionInner<P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    fn drop(&mut self) {
+        // Send the close signal to the client
+        match self.is_closed_sender.send(()) {
```

Comment:
impl drop for RpcClientConnectionInner  is not effected,  since RpcClientConnectionInner is hold by recv_loop. 
Delete the drop code,  move is_closed_sender to RpcClient,  receiver get notification when is_closed_sender is dropped.

Reply:
Ok, move sender to `RpcClient` and init inner with receiver.